### ISSUE TITLE
Add option for `v$` to exclude new line?

### ISSUE THREAD

**Issue by quicknir** [emacs-evil/evil#897]
# Issue type
- Enhancement request

# Environment

Emacs version: 25.1
Operating System: Name and version, e.g. Windows 10, macOS Sierra, Debian Jessie 
Evil version: evil-git-bff206a
Evil installation type: unsure (spacemacs)
Graphical/Terminal: graphical
Tested in a [clean environment](https://github.com/emacs-evil/evil/blob/master/CONTRIBUTING.md#bug-reports): No

# Reproduction steps
- Start Emacs
- Go to a line with some text
- try to do `y$` to copy some text, and paste
- it doesn't include newline
- try to do `v$y` to copy some text, and paste
- it does include newline

This isn't a bug report because vim behaves the same way. That said, I find this behavior to be unintuitive, surprising, and strictly less useful than if `v$` behaved consistently with `y$`. It would be amazing if there was an option added that enabled this. I would think that this shouldn't be too bad to implement, because the behavior of `$` is *currently* inconsistent depending on which operator it is used with. I would assume then that there must be some kind of special treatment of `$` when preceded by `v`. All that's needed is an option to disable that special treatment.

**Comment by wasamasa** [comment]
Isn't this about visual mode being inclusive?  Wouldn't something like `:set selection=exclusive` make more sense?

**Comment by quicknir** [comment]
@wasamasa Selection exclusive is an option with much more far reaching consequences, and having played with that option in vim that's definitely not what I want. And no, this is not about visual mode being inclusive. If you try this with any other motion, you will not observe the same thing. `yfx` and `vfxy` will yank exactly the same thing. The same is true for `w`, `e`, `t`, etc almost any motion. Why should `$` be special?

**Comment by quicknir** [comment]
Also as a follow up I've put this change into effect locally and it seems to work, I'll report here if I see any unexpected behavior.

**Comment by wasamasa** [comment]
Thanks, but no thanks.

**Comment by ninrod** [comment]
am I missing something? Why not just use `vg_`?

**Comment by quicknir** [comment]
@ninrod That's not the same thing, if there's trailing whitespace it will move before it. Also, `vg_` is not something I ever need to use, I don't see a reason to use more muscle memory on `vg_`. Flipping the question around: when is the last time you found `v$` grabbing the newline useful? I've never wanted it to do that so I find it a pure win. Obviously others may disagree, but that's why I proposed it as an option (and a non-default one at that).

**Comment by ninrod** [comment]
Well, I use `g_` all the time. But anyway, aren't you able to write a package to extend evil to perform as you mentioned for this case?

**Comment by braham-snyder** [comment]
@ninrod arguably this is analogous to https://github.com/emacs-evil/evil/issues/576 -- I have yet to test this particular change myself (I plan to), but the same argument might make sense here (i.e. vim seems inexplicably self-inconsistent, so it might be worth fixing it more directly in evil than in an entirely separate package)

**Comment by wasamasa** [comment]
It isn't really, the linked thing is a popular customization while this proposal is not.  Heck, it was discussed on r/vim before where the overall sentiment was split evenly between bewilderment at why you'd want this and the suggestion to just remap to a fancy keyboard macro.

**Comment by quicknir** [comment]
@wasamasa Yep, I'm the one who started that r/vim conversation because I was curious. That tends to be the 95% reaction to things in general when it comes to editor customizations: suggest fixes, or wonder why someone would want that.

Normally I'm reasonably sympathetic, but in this case: 1) it removes a useless inconsistency in vim, 2) it's useful for some people, and harmless for everyone else, 3) it's easy to change within evil and annoying to change outside of evil.

A couple of people have thumbs'ed up the original issue already, and it's not like this page has gotten a million views. For such a small, harmless, and internally consistent change, the bar of popularity should be pretty low.

**Comment by ninrod** [comment]
well, this is indeed an inconsistency of vim that can be remedied as an opt-in option.

if this behaviour is opt-in, does no harm and introduces no bugs, I don't see harm in merging this.

**Comment by Dickby** [comment]
what about `evil-want-visual-char-semi-exclusive` ?

**Comment by quicknir** [comment]
@Dickby As I've mentioned here, this isn't about visual mode being exclusive or not in general. This whole issue is entirely specific to $/the end of line motion/the newline character. So whatever the name is, it needs to reflect that one/several/all of those hthings. I think having $ in the name of the variable is the simplest way to reflect that, but any name that conveys that point is ok.

**Comment by Dickby** [comment]
I didn't came up with that name, the variable.

from the documentation:

> Visual character selection to beginning/end of line is exclusive.
If non nil then an inclusive visual character selection which
ends at the beginning or end of a line is turned into an
exclusive selection. Thus if the selected (inclusive) range ends
at the beginning of a line it is changed to not include the first
character of that line, and if the selected range ends at the end
of a line it is changed to not include the newline character of
that line.

**Comment by quicknir** [comment]
Oh, I thought you were proposing a name. Is this option mimicking something in vim, or what was the reasoning for this? In any case this doesn't do the same thing, and IMHO this option has some... weird effects. For example, if you highlight past the end of current line, to the first character on next line and delete, it's exclusive, so it doesn't take that character. But if you move the cursor one past, to the second character, it becomes inclusive again, so now the delete takes the second character. So, move the cursor by one, in visual mode, and then hitting delete, makes a two character difference...

I'd love to know the backstory of this option :-). I could ask the person who merged it why the heck they wanted it, or suggested a workaround...

**Comment by edkolev** [comment]
I see your point @quicknir, it does make sense. If it were up to me, I wouldn't accept this change since there are already 2 ways to do this, but that's just my opinion.

More importantly, however, I would kindly suggest you stop pushing this - maybe the change will be accepted, maybe it won't. My guess is pushing will only decrease the chances of it getting accepted - I know [Tim Pope](http://github.com/tpope) has intentionally decided not to fix particular bugs due to too many people +1-ing them.

In other words, not all the OSS maintainers have the luxury of being paid for what they do, let's not get them burned out.

**Comment by quicknir** [comment]
@edkolev That's surely not my intent to burn anyone out, but please be fair: I submitted a PR which I'd tested, and am more than willing to make any adjustments, write any tests, etc.

I just came here to mention another example where I think this option provides much more useful behavior. It's the kind of thing that you forget about, but drives you bonkers when you hit it. When I want to replace text with other text, my approach (and many people's) is to yank the text you want, then visually select the text you want to replace, and hit p, which replaces it. When you want to replace the whole end of the line, you yank something, do `v$p`, and then since `v$` grabs the newline character, your line gets joined to the next, which is rarely what you want. This really highlights the inconsistency, because often you may grab the text earlier with `Y` (changing the end of one line, to the end of another line). The `Y`/`y$` doesn't grab the new line, but `v$` does, so after the replacement you're short a new line and your lines have been joined.

**Comment by CeleritasCelery** [comment]
I would love to have this option. I always find my self using `v$` then `l` to correct it. Makes so much more sense.

**Comment by nixmaniack** [comment]
This is definitely something I'd like to have as an option. It's quite counter-intuitive to what `y$` does and it's a bit annoying to have to do `v$` and then `h` to get rid of the newline char. 

For me `v$` should mean `select till the end of line` and shouldn't include `new line`.

**Comment by vmandela** [comment]
Is there a way for me to add @quicknir 's fix from .emacs instead of editing evil source ? can the evil-end-of-line motion be redefined in .emacs?

**Comment by CeleritasCelery** [comment]
Yes. That’s the beauty of emacs. Just add it to your `.emacs` and restart. That’s what I have done.

**Comment by MoHKale** [comment]
Is there a reason this was never accepted? this being a really small change and completely optional, I don't really see why it was rejected. Having to basically redefine a function from evil in our config just to change one line to get the intended affect seems wasteful (for lack of a better word). Worse if that function definition ever changes outside of our config, we'll have to ensure its upto date less encounter some annoying bugs.

With the divide on whether people need this or not and the clear indication that a substantial amount do *want* it, wouldn't the best course of action be include one more customizable variable for this alongside the small alteration to evil-end-of-line?

**Comment by TheBB** [comment]
I guess the previous commenter deleted his comment. I can't speak for @wasamasa but if this is a popular option (which it seems to be) I'd be happy to merge it. On the other hand I'm assuming `evil-want-visual-char-semi-exclusive` was intended to solve this problem, and it definitely feels icky to have two options that do almost the same thing. :-(

**Comment by quicknir** [comment]
I'm not sure what problem that option was honestly trying to solve, I don't set that option and I can't imagine why I would. But I don't think it's solving the same problem. It applies to all visual ranges, regardless of how you got there, first of all, and it also causes a mismatch between the visual range and what actually gets taken. So it's more in the same vein as vim's inclusive/exclusive options. This is really an option for visual mode.

Despite misunderstandings to the contrary, the option I'm proposing has nothing to do with inclusive/exclusive. It's purely a "fix" to a *motion*, `$`. At the simplest level, for the vast majority of motions, they move the cursor to the same place regardless of whether you're in visual mode or not. `$` moves the cursor to a different place depending on whether visual mode is active or not. This behavior is silly, but it's vim's behavior, so it's reasonable for evil to mimic it by default, but it's also reasonable to have a way to turn it off.

**Comment by MoHKale** [comment]
oh, sorry for deleting it 😞. I saw `evil-want-visual-char-semi-exclusive` and assumed this was merged into evil in a different issue (explaining why I couldn't see the option in emacs and a merged notification here) and thus my comment was just plain wrong (I'll be more selective when removing comments in future, again apologies). In retrospect the two are different. I've set `evil-want-visual-char-semi-exclusive` to t in my config, and while $ doesn't include the whole end of the line in the visual selection it's distinct from $ in normal state with `evil-want-Y-yank-to-eol`. For example, if I've got the following text:

```elisp
  (add-hook 'prog-mode-hook #'my-super-cool-function-doing-super-cool-things)
```

if at any column in the line I hit $ my cursor/point is taken to the ending parenthesis. If I switch to visual state and do the same, I'm taken one character past the parenthesis just as I would've been had I not changed `evil-want-visual-char-semi-exclusive`. While performing any operations on the selection is still the same (if I yank it, eol is not included), if I did this operation so that I could return to the matching parenthesis, I still have to type `$h%` instead of `$%` as I would in normal state. It's a small difference but the idiosyncrasy/inconsistency behind it is annoying.

**Comment by mpereira** [comment]
Did this feature request ever land in evil-mode? I couldn't find `evil-v$-gets-eol` and `(setq evil-want-visual-char-semi-exclusive t)` doesn't seem to achieve having `v$` not selecting the newline after the last character in a line.

**Comment by ghbrown** [comment]
I would have preferred to use **quicknir**'s fix, but I didn't have any luck with it. Since I am not using general.el (employed by **mpereira**'s fix), I hacked this together:

```elisp
(define-key evil-visual-state-map (kbd "$")
  (lambda() (interactive)
    (evil-end-of-line)    ; to end of line (including newline)
    (evil-backward-char)  ; back to second to last visible character (weird)
    (evil-forward-char))) ; forward to last visible character
```
which seems to be working well so far, but I'll report back if it causes unintended side effects.

However, I'd much prefer an official fix like **quicknir**'s with an option to remedy this inconsistency in evil.

**Comment by tomdl89** [comment]
FWIW, I don't think `evil-want-visual-char-semi-exclusive` is particularly weird - just bad. It tries to solve the problem by:

- Excluding the `\n` when you move to the end of line (e.g. with `$`)
- Including the `\n` when you move one character further (i.e. the first char of the next line)

The second point was presumably introduced so the user has the option to press one extra char (`l`) to get the _inclusive_ behaviour, rather than the default situation, where the user has the option to press one extra char (`h`) to get the _exclusive_ behaviour. Unfortunately I don't think Frank considered (or cared) about the edge case where you want to select from a given point to the first character of the next line, inclusive. That's not possible with this defcustom set to non-nil, and I guess that's what people are suggesting is "weird".

I think `evil-want-visual-char-semi-inclusive` is poorly thought out, and should be deprecated. It strikes me that simply adding an option to change the behaviour of `evil-end-of-line` as @quicknir suggested better addresses the issue, without introducing any weird edge case.
